xdsl/irdl/constraints.py

```diff
@@ -165,12 +165,10 @@ def __or__(
             return value  # pyright: ignore[reportReturnType]
         return AnyOf((self, value))
 
-    def __and__(
-        self, value: AttrConstraint[AttributeCovT], /
-    ) -> AttrConstraint[AttributeCovT]:
+    def __and__(self, value: AttrConstraint, /) -> AttrConstraint[AttributeCovT]:
```

Comment:
In the absense of an intersection type, I propose just taking the first type to be the bound, as opposed to all of the constraints needing to be of the same type